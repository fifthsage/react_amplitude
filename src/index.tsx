import React, { useState, useEffect, useContext, createContext } from "react";

let amplitude = null;

if (typeof window !== "undefined") {
  amplitude = require("amplitude-js/amplitude");
}

const instance = amplitude ? amplitude.getInstance() : null;

interface AmplitudeProviderProps {
  logEvent: Function;
  event: LogEventProps | null;
}

interface LogEventProps {
  key: string | null;
  payload: { [key: string]: any };
}

const amplitudeContext = createContext<AmplitudeProviderProps>({
  logEvent: Function,
  event: null
});

function useProvideAmplitude(): AmplitudeProviderProps {
  const [event, setEvent] = useState<LogEventProps>({
    key: null,
    payload: {}
  });

  const logEvent = (key: string, payload: { [key: string]: any }) => {
    if (instance) {
      instance.logEvent(key, payload);

      setEvent({
        key: key,
        payload: payload
      });
    }

    return;
  };

  useEffect(() => {
    return;
  }, []);

  return {
    event,
    logEvent
  };
}

export const ProvideAmplitude = (props: any) => {
  const {
    apiKey,
    userId = null,
    config = {},
    disabled = false,
    children
  } = props;

  const defaultOptions = {
    saveEvents: true,
    includeUtm: true,
    includeReferrer: true,
    trackingOptions: {
      city: false,
      ip_address: false
    }
  };

  if (instance && !disabled) {
    instance.init(apiKey, userId, Object.assign(defaultOptions, config));
  }

  const _amplitude = useProvideAmplitude();

  return (
    <amplitudeContext.Provider value={_amplitude}>
      {children}
    </amplitudeContext.Provider>
  );
};

export const ConsumeAmplitude = amplitudeContext.Consumer;

export const useAmplitude = () => {
  return useContext(amplitudeContext);
};
