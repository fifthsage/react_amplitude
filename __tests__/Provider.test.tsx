import React, { FunctionComponent } from "react";
import { ProvideAmplitude, useAmplitude } from "../src";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";

// jest.mock("amplitude-js");

const apiKey = "a41e51ccfa11b3a8c9667d0c2eee620e";

const LogTestWrapper: FunctionComponent = () => {
  const amplitudeEvent = useAmplitude();

  return (
    <div>
      <div>{(amplitudeEvent.event || {}).toString()}</div>
      <button
        onClick={() => amplitudeEvent.logEvent("TEST_EVENT", { key: "value" })}
      >
        send
      </button>
    </div>
  );
};

describe("suite", () => {
  it("should render", () => {
    const component = shallow(
      <ProvideAmplitude apiKey={apiKey}></ProvideAmplitude>
    );

    expect(component.length).toBe(1);

    expect(toJson(component)).toMatchSnapshot();
  });

  it("should render with Log Test", () => {
    const component = mount(
      <ProvideAmplitude apiKey={apiKey}>
        <LogTestWrapper />
      </ProvideAmplitude>
    );

    expect(component.length).toBe(1);
    expect(component.find(LogTestWrapper).length).toBe(1);
    component.find("button").simulate("click");

    expect(toJson(component)).toMatchSnapshot();
  });
});
