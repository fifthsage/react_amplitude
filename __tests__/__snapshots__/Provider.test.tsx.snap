// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`suite should render 1`] = `
<ContextProvider
  value={
    Object {
      "event": Object {
        "key": null,
        "payload": Object {},
      },
      "logEvent": [Function],
    }
  }
/>
`;

exports[`suite should send event with Consumer 1`] = `
<ProvideAmplitude
  apiKey="a41e51ccfa11b3a8c9667d0c2eee620e"
>
  <button
    onClick={[Function]}
  >
    send
  </button>
</ProvideAmplitude>
`;

exports[`suite should send event with useContext 1`] = `
<ProvideAmplitude
  apiKey="a41e51ccfa11b3a8c9667d0c2eee620e"
>
  <LogTestWrapper>
    <div>
      <div>
        [object Object]
      </div>
      <button
        onClick={[Function]}
      >
        send
      </button>
    </div>
  </LogTestWrapper>
</ProvideAmplitude>
`;
